### PR TITLE
[#451] duplicate plan name validation

### LINKED ISSUE #451: Duplicate Plan Names
I know the term "duplicate" is starting to make us shriek...

Per our last email on this subject - we would like to use the UI to work around the duplicate plan name issue in the short term by simply checking the service template (plans) in memory (loaded from Overview) and passing them into the Plan Wizard (to allow checking for duplicates). This is a short term fix - the long term fix is obviously adding this validation on the backend.

Would like to resolve this for 5.9.4.

BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1567225

### PR DESCRIPTION
**Pull request by priley86**
Fixes #451 
Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1567225

* adds the duplicate plan name validation to the Plan Wizard (similar to duplicate plan name in Mapping Wizard)
* Passes in all plans loaded from Overview to the Plan Wizard - *note*: we will need to query archived plans on load of Overview now to support this (they will be loaded in the background though).
* adds the Redux alert actions like we have in the Mapping Wizard

plan-name-validation

### CODE REVIEW

**Comment by priley86** [comment]
@michaelkro the user is unable to click Next since we've prevented it in the `nextStep` logic, but I agree we should also probably disable the Next button in that case... this is consistent w/ the Infra Mapping Wizard validation so we should probably update both to handle that consistently... I can tweak this 😸 

screen shot 2018-07-10 at 8 39 54 am


screen shot 2018-07-10 at 8 41 03 am

**Comment by priley86** [comment]
disabled next step button if asyncErrors for any form...

File: app/javascript/react/screens/App/Overview/screens/PlanWizard/components/PlanWizardGeneralStep/helpers.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+      (names, plan) => [...names, plan.name.trim()],
+      []
+    );
+    const existingArchivedPlanNames = archivedTransformationPlans.reduce(
+      (names, plan) => [...names, plan.name.trim()],
+      []
+    );
+
+    const allPlanNames = [...existingTransformationPlanNames, ...existingArchivedPlanNames];
+    const duplicateName = allPlanNames.find(existingPlanName => existingPlanName === newPlanName.trim());
```
Comment (AllenBW) [inline app/javascript/react/screens/App/Overview/screens/PlanWizard/components/PlanWizardGeneralStep/helpers.js]: So we trim whitespace, but should "VmWithAnAwesomeName" be considered the same thing as "vmwithanawesomename"?  Should name duplication gonna be case sensitive?

Comment (michaelkro) [reply app/javascript/react/screens/App/Overview/screens/PlanWizard/components/PlanWizardGeneralStep/helpers.js]: I think the back end does not validate for case sensitivity.  At least this was the case when we were looking at unique mapping names

File: app/javascript/react/screens/App/Overview/Overview.js [hunk 2]
```diff
@@ -32,6 +32,7 @@ class Overview extends React.Component {
       fetchTransformationMappingsAction,
       fetchTransformationPlansUrl,
       fetchTransformationPlansAction,
+      fetchArchivedTransformationPlansUrl,
```
Comment (AparnaKarve) [inline app/javascript/react/screens/App/Overview/Overview.js]: Nice! The duplication check is across all Plans - Archived and Active.

**Comment by priley86** [comment]
cool - thanks for the reviews all 👍

**Comment by simaishi** [comment]
Gaprindashvili backport details:

```
$ git log -1
commit 594bd65a458d4482d0d5fec74c32564c9a1d8de2
Author: Aparna Karve <aparna.karve@gmail.com>
Date:   Tue Jul 10 09:43:17 2018 -0700

    Merge pull request #474 from priley86/duplicate-plans
    
    [#451] duplicate plan name validation
    (cherry picked from commit 8bf1179dbdeea1e7de83ae7eb727bbc8e5ef6d9b)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1608768
```

**Comment by simaishi** [comment]
@priley86 This is causing Travis failure in Gaprindashvili branch, please take a look:
https://travis-ci.org/ManageIQ/miq_v2v_ui_plugin/builds/408443694

```
/home/travis/build/ManageIQ/miq_v2v_ui_plugin/app/javascript/react/screens/App/Overview/screens/PlanWizard/components/PlanWizardGeneralStep/index.js
  17:24  error  Replace `⏎··mapStateToProps,⏎··actions⏎` with `mapStateToProps,·actions`  prettier/prettier
```

**Comment by vconzola** [comment]
@simaishi @priley is out for the next two weeks - he's a new father. @AparnaKarve can you take look at this?

**Comment by AparnaKarve** [comment]
@simaishi https://github.com/ManageIQ/miq_v2v_ui_plugin/pull/519 fixes the test

**Pull request merged**